src/org/intellij/erlang/quickfixes/ErlangCreateFunctionQuickFix.java

```diff
@@ -24,81 +25,104 @@
+    super(fixMessage, "Erlang");
+    myFunctionText = new FunctionTextProvider<PsiElement>(element, name, arity);
+    myFixMessage = fixMessage;
+  }
+
+  public ErlangCreateFunctionQuickFix(@NotNull FunctionTextProvider<? extends PsiElement> provider,
+                                      @NotNull String fixMessage) {
+    super(fixMessage, "Erlang");
+    myFunctionText = provider;
+    myFixMessage = fixMessage;
   }
 
   @Override
-  public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
-    if (myArity < 0) return;
-    //noinspection unchecked
-    PsiElement call = PsiTreeUtil.getContextOfType(
-      descriptor.getPsiElement(),
-      false,
-      ErlangFunctionCallExpression.class,
-      ErlangSpecFun.class,
-      ErlangFunctionWithArity.class,
-      ErlangExportFunction.class
-    );
-    if (call != null) {
-      //noinspection unchecked
-      PsiElement topmost = PsiTreeUtil.getParentOfType(call,
-        ErlangFunction.class, ErlangRecordDefinition.class, ErlangMacros.class, ErlangAttribute.class, ErlangModule.class);
-
-      if (call instanceof ErlangExportFunction) {
-        topmost = call.getContainingFile().getLastChild();
-      }
+  public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor problemDescriptor) {
+  }
+
+  @Nls
+  @NotNull
+  @Override
+  public String getText() {
+    return myFixMessage;
+  }
+
+  @Override
+  public boolean isAvailable(@NotNull Project project, Editor editor, PsiFile file) {
+    return true;
+  }
+
+  @Override
+  public void invoke(@NotNull Project project, Editor editor, PsiFile file) throws IncorrectOperationException {
+    int textOffset = file.getLastChild().getTextRange().getEndOffset();
+    TemplateManager templateManager = TemplateManager.getInstance(project);
+    Template template = templateManager.createTemplate("", "");
+    template.setToReformat(true);
+    template.addTextSegment("\n\n");
+    template.addTextSegment(myFunctionText.getName() + "(");
+
+    List<String> arguments = myFunctionText.getArguments();
+    int arity = arguments.size();
+    for (int i = 0; i < arity; i++) {
+      String name = arguments.get(i);
+      template.addVariable("param" + i, new ConstantNode(name), true);
+      if (i != arity - 1) template.addTextSegment(", ");
+    }
+    template.addTextSegment(") ->\n");
+    template.addEndVariable();
+    template.addTextSegment("erlang:error(not_implemented).");
+
+    editor.getCaretModel().moveToOffset(textOffset);
+    templateManager.startTemplate(editor, template);
+  }
+
+  @Override
+  public boolean startInWriteAction() {
+    return true;
+  }
+
+  public static class FunctionTextProvider<T extends PsiElement> {
+    private final String myName;
+    private final int myArity;
+    protected final SmartPsiElementPointer<T> myPointerToProblemPsiElement;
+
+    public FunctionTextProvider(@NotNull T element, @NotNull String name, int arity) {
```

Comment:
And, we're back where we started from. We've spent a couple of days to understand we don't need to store name and arity, and we store it again...